// @foal/storage/lib/index.js
"use strict";
/**
 * FoalTS
 * Copyright(c) 2017-2022 Loïc Poullain <loic.poullain@centraliens.net>
 * Released under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidateMultipartFormDataBody = exports.LocalDisk = exports.File = exports.isFileDoesNotExist = exports.FileDoesNotExist = exports.Disk = void 0;
var disk_service_1 = require("./disk.service");
Object.defineProperty(exports, "Disk", { enumerable: true, get: function () { return disk_service_1.Disk; } });
Object.defineProperty(exports, "FileDoesNotExist", { enumerable: true, get: function () { return disk_service_1.FileDoesNotExist; } });
Object.defineProperty(exports, "isFileDoesNotExist", { enumerable: true, get: function () { return disk_service_1.isFileDoesNotExist; } });
var file_1 = require("./file");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return file_1.File; } });
var local_disk_service_1 = require("./local-disk.service");
Object.defineProperty(exports, "LocalDisk", { enumerable: true, get: function () { return local_disk_service_1.LocalDisk; } });
var validate_multipart_form_data_body_hook_1 = require("./validate-multipart-form-data-body.hook");
Object.defineProperty(exports, "ValidateMultipartFormDataBody", { enumerable: true, get: function () { return validate_multipart_form_data_body_hook_1.ValidateMultipartFormDataBody; } });
