// @foal/core/lib/common/hooks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidateQueryParam = exports.ValidatePathParam = exports.ValidateHeader = exports.ValidateCookie = exports.ValidateBody = exports.UserRequired = exports.Log = void 0;
var log_hook_1 = require("./log.hook");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return log_hook_1.Log; } });
var user_required_hook_1 = require("./user-required.hook");
Object.defineProperty(exports, "UserRequired", { enumerable: true, get: function () { return user_required_hook_1.UserRequired; } });
var validate_body_hook_1 = require("./validate-body.hook");
Object.defineProperty(exports, "ValidateBody", { enumerable: true, get: function () { return validate_body_hook_1.ValidateBody; } });
var validate_cookie_hook_1 = require("./validate-cookie.hook");
Object.defineProperty(exports, "ValidateCookie", { enumerable: true, get: function () { return validate_cookie_hook_1.ValidateCookie; } });
var validate_header_hook_1 = require("./validate-header.hook");
Object.defineProperty(exports, "ValidateHeader", { enumerable: true, get: function () { return validate_header_hook_1.ValidateHeader; } });
var validate_path_param_hook_1 = require("./validate-path-param.hook");
Object.defineProperty(exports, "ValidatePathParam", { enumerable: true, get: function () { return validate_path_param_hook_1.ValidatePathParam; } });
var validate_query_param_hook_1 = require("./validate-query-param.hook");
Object.defineProperty(exports, "ValidateQueryParam", { enumerable: true, get: function () { return validate_query_param_hook_1.ValidateQueryParam; } });
