// @foal/core@2.9.0/lib/common/utils/get-ajv-instance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAjvInstance = exports._instanceWrapper = void 0;
// 3p
const Ajv = require("ajv");
const core_1 = require("../../core");
// This is a little hack to test the customized configuration of `getAjvInstance`.
// tslint:disable-next-line:variable-name
exports._instanceWrapper = {
    instance: null
};
/**
 * Return the Ajv instance used internally by FoalTS validation hooks.
 *
 * It has this default configuration:
 *  - coerceTypes: true (Change data type of data to match `type` keyword.)
 *  - removeAdditional: true (Remove additional properties when `additionalProperties` keyword is false.)
 *  - useDefaults: true (Replace missing properties and items with the values from corresponding `default` keyword)
 *
 *
 * @export
 * @returns {Ajv.Ajv} The AJV instance
 */
function getAjvInstance() {
    if (!exports._instanceWrapper.instance) {
        exports._instanceWrapper.instance = new Ajv({
            $data: core_1.Config.get('settings.ajv.$data', 'boolean'),
            allErrors: core_1.Config.get('settings.ajv.allErrors', 'boolean'),
            coerceTypes: core_1.Config.get('settings.ajv.coerceTypes', 'boolean|string', true),
            nullable: core_1.Config.get('settings.ajv.nullable', 'boolean'),
            removeAdditional: core_1.Config.get('settings.ajv.removeAdditional', 'boolean|string', true),
            useDefaults: core_1.Config.get('settings.ajv.useDefaults', 'boolean|string', true),
        });
    }
    return exports._instanceWrapper.instance;
}
exports.getAjvInstance = getAjvInstance;
